src/Common/Core/Impl/OS/Win32EnvironmentBlock.cs

```diff
@@ -0,0 +1,102 @@
+            Win32EnvironmentBlock eb = new Win32EnvironmentBlock();
+            try {
+                if (CreateEnvironmentBlock(out env, token, inherit)) {
+                    IntPtr ptr = env;
+                    while (true) {
+                        string envVar = Marshal.PtrToStringUni(ptr);
+                        byte[] data = Encoding.Unicode.GetBytes(envVar);
+                        if (data.Length <= 2) {
+                            // detected double null
+                            break;
+                        } else {
+                            int idx = envVar.IndexOf('=');
+                            if (idx > 0 && idx < envVar.Length) {
+                                string key = envVar.Substring(0, idx);
+                                string value = envVar.Substring(idx + 1);
+                                eb.Add(key, value);
+                            }
+                        }
+                        // unicode string + unicode null
+                        ptr += data.Length + 2;
+                    }
+                }
+            } finally {
+                DestroyEnvironmentBlock(env);
+            }
+            return eb;
+        }
+
+        public byte[] ToByteArray() {
+            using (MemoryStream ms = new MemoryStream()) {
+                byte[] nulls = { 0, 0 };
+                foreach (var p in this) {
+                    string envData = $"{p.Key}={p.Value}";
+                    byte[] data = Encoding.Unicode.GetBytes(envData);
+                    ms.Write(data, 0, data.Length);
+                    ms.Write(nulls, 0, nulls.Length);
+                }
+                ms.Write(nulls, 0, nulls.Length);
+                return ms.ToArray();
+            }
+        }
+
+        public Win32NativeEnvironmentBlock GetNativeEnvironmentBlock() {
+            int length;
+            return GetNativeEnvironmentBlock(out length);
+        }
+
+        public Win32NativeEnvironmentBlock GetNativeEnvironmentBlock(out int length) {
+            byte[] data = ToByteArray();
+            IntPtr ptr = Marshal.AllocHGlobal(data.Length);
+
+            Marshal.Copy(data, 0, ptr, data.Length);
+
+            length = data.Length;
+            return new Win32NativeEnvironmentBlock(ptr);
+        }
+
+        public new string this[string key] {
+            get {
+                string mapped = GetMappedKey(key);
+                if (string.IsNullOrWhiteSpace(mapped)) {
+                    return null; 
+                } else {
+                    return base[mapped];
+                }
+            }
+            set {
+                string mapped = GetMappedKey(key);
+                if (!string.IsNullOrWhiteSpace(mapped)) {
+                    base[mapped] = value;
+                } else {
+                    Add(key, value);
+                }
+            }
+        }
+
+        private string GetMappedKey(string key) {
+            string low = key.ToLowerInvariant();
```

Comment:
This mapping was needed to maintain the original casing of the environment block. TryGetValue does not seem to have a way to set the string comparer. If we don't use this mapping we can create, for example, "PATH" and "Path" as environment variables, while we add the interpreter path. Depending on which evnvariable windows picks-up, LoadLibrary can file while loading RHost since it cannot find the R.dll